examples/modular-drivers/src/bme280_ulp.rs

```diff
@@ -0,0 +1,970 @@
+                .ok_or(BME280Error::BME280UlpInvalidRTCPin(cfg.pin_sda))
+        }? as u32;
+
+        let ulp = unsafe { micro_rdk::esp32::esp_idf_svc::hal::ulp::ULP::new() };
+        let mut ulp = UlpDriver::new(ulp).map_err(BME280Error::BME280UlpDriverError)?;
+        unsafe { ulp.load(ULP_PROGRAM) }.map_err(BME280Error::BME280UlpLoadError)?;
+
+        Ok(Self {
+            ulp,
+            rtc_pin_scl,
+            rtc_pin_sda,
+            cfg,
+        })
+    }
+    // 31      28 27   23 22   18 17     9
+    // | OpCode |  High |  Low |  Unused | Addr
+    // ESP32 Technical reference manual 30.4.13
+    fn ulp_fsm_reg_rd_inst(rtc_reg: u32, low_bit: u32, bit_width: u32) -> u32 {
+        use micro_rdk::esp32::esp_idf_svc::sys::DR_REG_RTCCNTL_BASE;
+        (2_u32 << 28)
+            | (((low_bit + bit_width - 1) & ((1 << 5) - 1)) << 23)
+            | (((low_bit) & ((1 << 5) - 1)) << 18)
+            | ((rtc_reg - DR_REG_RTCCNTL_BASE) / 4)
+    }
+    // 31      28 27   23 22   18 17  10 9
+    // | OpCode |  High |  Low |  Data  | Addr
+    // ESP32 Technical reference manual 30.4.14
+    fn ulp_fsm_reg_wr_inst(rtc_reg: u32, low_bit: u32, bit_width: u32, data: u32) -> u32 {
+        use micro_rdk::esp32::esp_idf_svc::sys::DR_REG_RTCCNTL_BASE;
+        (1_u32 << 28)
+            | (((low_bit + bit_width - 1) & ((1 << 5) - 1)) << 23)
+            | (((low_bit) & ((1 << 5) - 1)) << 18)
+            | (((data) & ((1 << 8) - 1)) << 10)
+            | ((rtc_reg - DR_REG_RTCCNTL_BASE) / 4)
+    }
+
+    fn start_ulp(&mut self, mut cfg: Configuration) -> Result<(), BME280Error> {
+        use micro_rdk::esp32::esp_idf_svc::sys::{
+            esp, rtc_gpio_init,
+            rtc_gpio_mode_t_RTC_GPIO_MODE_INPUT_OUTPUT as RTC_GPIO_MODE_INPUT_OUTPUT,
+            rtc_gpio_set_direction, ulp_set_wakeup_period, RTC_GPIO_ENABLE_W1TC_REG,
+            RTC_GPIO_ENABLE_W1TC_S, RTC_GPIO_ENABLE_W1TS_REG, RTC_GPIO_ENABLE_W1TS_S,
+            RTC_GPIO_IN_NEXT_S, RTC_GPIO_IN_REG,
+        };
+        // configure pin so they can be accessed by ULP. Note that any pin should not be changed or used afterwards
+        esp!(unsafe { rtc_gpio_init(self.cfg.pin_scl) })
+            .and_then(|_| {
+                esp!(unsafe {
+                    rtc_gpio_set_direction(self.cfg.pin_scl, RTC_GPIO_MODE_INPUT_OUTPUT)
+                })
+            })
+            .map_err(|err| BME280Error::BME280UlpRtcPinConfigError(self.cfg.pin_scl, err))?;
+        esp!(unsafe { rtc_gpio_init(self.cfg.pin_sda) })
+            .and_then(|_| {
+                esp!(unsafe {
+                    rtc_gpio_set_direction(self.cfg.pin_sda, RTC_GPIO_MODE_INPUT_OUTPUT)
+                })
+            })
+            .map_err(|err| BME280Error::BME280UlpRtcPinConfigError(self.cfg.pin_sda, err))?;
+
+        // config NOP reserved insts
+        unsafe {
+            *set_SCL = Self::ulp_fsm_reg_wr_inst(
```

Comment:
so `i2c.S` (and others) are globals (whose values represent registers used by the ULP program) that you're populating here?

Reply:
we are replacing NOP instruction with their respective REG_RD/REG_WR instructions. This portion of code is link the the data section (RW) and not the text section (normally RO)